src/sql/src/plan/expr.rs

```diff
@@ -1386,6 +1523,584 @@ impl HirRelationExpr {
     }
 }
 
+impl VisitChildren<Self> for HirRelationExpr {
+    fn visit_children<F>(&self, mut f: F)
+    where
+        F: FnMut(&Self),
+    {
+        // subqueries of type HirRelationExpr might be wrapped in
+        // Exists or Select variants within HirScalarExpr trees
+        // attached at the current node, and we want to visit them as well
+        VisitChildren::visit_children(self, |expr: &HirScalarExpr| {
+            #[allow(deprecated)]
+            Visit::visit_post_nolimit(expr, &mut |expr| match expr {
+                HirScalarExpr::Exists(expr) | HirScalarExpr::Select(expr) => f(expr.as_ref()),
+                _ => (),
+            });
```

Comment:
What are your thoughts about moving this closure body into an `imp VisitChildren<HirRelationExpr> for HirScalarExpr`, for symmetry?

Reply:
I have some doubts whether the a symmetric definition will preserve the semantics of the `Visit::*visit*` methods with their current definitions. I wanted to write down a simple test to validate that every child is visited only once in recursive traversals if we have symmetric definitions of `VisitChildren<A> for A` and ``VisitChildren<B> for B` that "peek through" the corresponding type.

Reply:
Sadly as I anticipated that didn't work. I added a test in `visit.rs`, and you can check for yourself that the `test_recursive_types_b` test will not even compile if we make the definitions symmetric. Look at the comment on the test for more details. 

Certainly this is a shortcoming of the current approach, and in the long term I think the better solution for the framework is to just generalize the callbacks so they can accept arguments from different types in the case of a mutually recursive data type.

While limited, however, this variant is sufficient to cover our immediate needs, so I will merge it and punt the discussion of how to adequately generalize this for later.